### PR TITLE
[bridge 14/n] simple handle_eth_tx_hash

### PR DESCRIPTION
**Pull request by longbowlu**
## Description 

1. implement `get_finalized_bridge_action_maybe` in `EthClient`
2. implement simple logic for `BridgeRequestHandler::handle_eth_tx_hash`. I'm looking to add caching + NotifyRead-ish thing in a future PR.

## Test Plan 

unit tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
implement `get_finalized_bridge_action_maybe` in `EthClient` and implement simple logic for `BridgeRequestHandler::handle_eth_tx_hash`

### CODE REVIEW

File: crates/sui-bridge/src/test_utils.rs [hunk 1]
```diff
@@ -108,3 +116,93 @@ pub fn sign_action_with_key(
     let sig = BridgeAuthoritySignInfo::new(action, secret);
     SignedBridgeAction::new_from_data_and_sig(action.clone(), sig)
 }
+
+pub fn mock_last_finalized_block(mock_provider: &EthMockProvider, block_number: u64) {
```
Comment (longbowlu) [inline crates/sui-bridge/src/test_utils.rs:120]: moved from other files for code sharing, can skip

**Comment by longbowlu** [comment]
cc @Jayfromthe13th

File: crates/sui-bridge/src/eth_client.rs [hunk 2]
```diff
@@ -58,11 +58,39 @@ where
...
+            .provider
+            .get_transaction_receipt(tx_hash)
+            .await
+            .map_err(BridgeError::from)?
+            .ok_or(BridgeError::TxNotFound)?;
+        let receipt_block_num = receipt.block_number.ok_or(BridgeError::ProviderError(
+            "Provider returns log without block_number".into(),
+        ))?;
+        let last_finalzied_block_id = self.get_last_finalized_block_id().await?;
+        if receipt_block_num.as_u64() > last_finalzied_block_id {
```
Comment (arun-koshy) [inline crates/sui-bridge/src/eth_client.rs:76]: This line would only be true for the period in between the block proposal and until the block receives enough votes. But in the case of network delays or equivocation by validators, where the blockchain's state might temporarily diverge, will we need some additional checks?

https://ethereum.org/en/developers/docs/consensus-mechanisms/pos/#fork-choice

File: crates/sui-bridge/src/error.rs [hunk 3]
```diff
@@ -11,8 +11,12 @@ pub enum BridgeError {
...
     // The referenced transction does not exist
     TxNotFound,
-    // The referenced transaction does not contain bridge events
-    NoBridgeEventsInTx,
+    // Tx is not yet finalized
+    TxNotFinalized,
+    // No recognized bridge event in specified transaction and event position
+    NoBridgeEventsInTxPosition,
+    // Found BridgeEvent but not BridgeAction
+    BridgeEventNotAction,
```
[the remark this answers is not in the document]
Comment (longbowlu) [reply crates/sui-bridge/src/error.rs:19]: `BridgeEventNoActionFound` ?

File: crates/sui-bridge/src/server/handler.rs [hunk 4]
```diff
@@ -46,10 +51,20 @@ impl BridgeRequestHandler {
 impl BridgeRequestHandlerTrait for BridgeRequestHandler {
     async fn handle_eth_tx_hash(
         &self,
-        _tx_hash_hex: String,
-        _event_idx: u16,
+        tx_hash_hex: String,
+        event_idx: u16,
     ) -> Result<Json<SignedBridgeAction>, BridgeError> {
-        unimplemented!()
+        // TODO add caching and avoid simalutaneous requests
```
[the remark this answers is not in the document]
Comment (longbowlu) [reply crates/sui-bridge/src/server/handler.rs:57]: only performance issue

File: crates/sui-bridge/src/eth_client.rs (hunk 2 above)
Comment (longbowlu) [reply crates/sui-bridge/src/eth_client.rs:76]: "finalized" blocks are considered highly unlikely to be reverted in forks. This simplifies the model the code is written. 

> A finalized block is a justified block that is 1 epoch behind the most recently justified block. Finalized blocks are extremely [unlikely to be re-organized](https://blog.ethereum.org/2021/11/29/how-the-merge-impacts-app-layer), with the only exception being if a two-thirds majority of validators finalize a competing chain of blocks.

https://www.alchemy.com/overviews/ethereum-commitment-levels

But in the event of an actual fork involving a previous finalized block, the code cannot handle it properly ...

File: crates/sui-bridge/src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (longbowlu) [reply crates/sui-bridge/src/error.rs:19]: i renamed it to `BridgeEventNotActionable`

File: crates/sui-bridge/src/eth_client.rs [hunk 5]
```diff
@@ -58,11 +58,39 @@ where
...
+        let receipt = self
+            .provider
+            .get_transaction_receipt(tx_hash)
+            .await
+            .map_err(BridgeError::from)?
+            .ok_or(BridgeError::TxNotFound)?;
+        let receipt_block_num = receipt.block_number.ok_or(BridgeError::ProviderError(
+            "Provider returns log without block_number".into(),
+        ))?;
+        let last_finalzied_block_id = self.get_last_finalized_block_id().await?;
```
Comment (arun-koshy) [inline crates/sui-bridge/src/eth_client.rs:75]: s/last_finalzied_block_id/last_finalized_block_id

**Review by arun-koshy — APPROVED** [review]

**Pull request merged**